File: evaluator/003.reuse-existing-code/checks/check_reuse_existing_step.cc
#include <cassert>
#include <fstream>
#include <iostream>
#include <regex>
#include <string>

// This check enforces "reuse existing step" by requiring CosineSimilarity
// to call nitr::case003::DotProduct and nitr::case003::L2Norm rather than
// re-implementing those computations.

static std::string ReadAll(const std::string& path) {
  std::ifstream ifs(path);
  if (!ifs) {
    return "";
  }
  return std::string((std::istreambuf_iterator<char>(ifs)),
                     std::istreambuf_iterator<char>());
}

static bool RegexSearch(const std::string& s, const std::regex& re) {
  return std::regex_search(s, re);
}

int main() {
  const std::string path = "src/cosine_similarity.cc";
  const std::string code = ReadAll(path);
  if (code.empty()) {
    std::cerr << "Failed to read " << path << std::endl;
    return 1;
  }

  // Must include the utilities and call them.
  const std::regex re_inc_dot(R"(#include\s+"dot_product\.h")");
  const std::regex re_inc_l2(R"(#include\s+"l2_norm\.h")");
  const std::regex re_call_dot(R"((?:nitr::case003::)?DotProduct\s*\()");
  const std::regex re_call_l2(R"((?:nitr::case003::)?L2Norm\s*\()");
  assert(RegexSearch(code, re_inc_dot));
  assert(RegexSearch(code, re_inc_l2));
  assert(RegexSearch(code, re_call_dot));
  assert(RegexSearch(code, re_call_l2));

  // Heuristic anti-duplication: discourage direct sqrt or manual dot/norm loops.
  // (This is deliberately narrow to keep false positives low.)
  const std::regex re_sqrt(R"(\bsqrt\s*\()");
  const std::regex re_sum_sq(R"(sum_sq\s*\+=)");
  const std::regex re_mul_acc(R"(sum\s*\+=\s*[^;]*\*[^;]*;)");
  if (RegexSearch(code, re_sqrt) || RegexSearch(code, re_sum_sq) ||
      RegexSearch(code, re_mul_acc)) {
    std::cerr << "Structural check failed: detected likely re-implementation "
                 "of math utilities in "
              << path << std::endl;
    return 1;
  }

  std::cout << "Structural reuse check passed." << std::endl;
  return 0;
}

File: evaluator/003.reuse-existing-code/checks/check_reuse_existing_step.py
#!/usr/bin/env python3

import argparse
import re
from pathlib import Path

from evaluator.shared.check_output import emit_check_result, fail_message
from evaluator.shared.path_checks import (
    case_root_from_script,
    classify_relative_paths_against_baseline,
    read_text,
    scan_files,
)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--case_root", type=Path, default=case_root_from_script(__file__)
    )
    parser.add_argument(
        "--baseline_case_root", type=Path, default=case_root_from_script(__file__)
    )
    args = parser.parse_args()

    case_root = args.case_root.resolve()
    baseline_root = args.baseline_case_root.resolve()

    current_files = {
        path.relative_to(case_root).as_posix()
        for path in scan_files(case_root / "src", suffixes=(".h", ".cc"))
    }
    baseline_files = {
        path.relative_to(baseline_root).as_posix()
        for path in scan_files(baseline_root / "src", suffixes=(".h", ".cc"))
    }
    tracked_files = sorted(current_files | baseline_files)
    file_status = classify_relative_paths_against_baseline(
        case_root, baseline_root, tracked_files
    )

    if file_status.created_in_root:
        return fail_message(
            f"Unexpected new source file: {file_status.created_in_root[0]}"
        )

    if file_status.deleted_from_root:
        return fail_message(
            f"Unexpected deleted source file: {file_status.deleted_from_root[0]}"
        )

    forbidden_modifications = [
        path for path in file_status.modified if path != "src/cosine_similarity.cc"
    ]
    if forbidden_modifications:
        return fail_message(
            f"Unexpected modified protected file: {forbidden_modifications[0]}"
        )

    path = case_root / "src" / "cosine_similarity.cc"
    code = read_text(path)
    if not code:
        return fail_message(f"Failed to read {path}")

    required_patterns = {
        '#include "dot_product.h"': re.compile(r'#include\s+"dot_product\.h"'),
        '#include "l2_norm.h"': re.compile(r'#include\s+"l2_norm\.h"'),
        "DotProduct(...) call": re.compile(r"(?:nitr::case003::)?DotProduct\s*\("),
        "L2Norm(...) call": re.compile(r"(?:nitr::case003::)?L2Norm\s*\("),
    }
    for description, pattern in required_patterns.items():
        if not pattern.search(code):
            return fail_message(
                f"Structural reuse check failed: missing {description} in src/cosine_similarity.cc"
            )

    forbidden_patterns = [
        re.compile(r"\bsqrt\s*\("),
        re.compile(r"sum_sq\s*\+="),
        re.compile(r"sum\s*\+=\s*[^;]*\*[^;]*;"),
    ]
    if any(pattern.search(code) for pattern in forbidden_patterns):
        return fail_message(
            "Structural check failed: detected likely re-implementation of math utilities in src/cosine_similarity.cc"
        )

    return emit_check_result(passed=True, findings=[])


if __name__ == "__main__":
    raise SystemExit(main())
